<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Highly variable gene selection methods, including popular public available methods, and also the mixture of multiple highly variable gene selection methods, &lt;https://github.com/RuzhangZhao/mixhvg&gt;. Reference: &lt;doi:10.1101/2024.08.25.608519&gt;.">
<title>Mixture of Multiple Highly Variable Feature Selection Methods • mixhvg</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Mixture of Multiple Highly Variable Feature Selection Methods">
<meta property="og:description" content="Highly variable gene selection methods, including popular public available methods, and also the mixture of multiple highly variable gene selection methods, &lt;https://github.com/RuzhangZhao/mixhvg&gt;. Reference: &lt;doi:10.1101/2024.08.25.608519&gt;.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">mixhvg</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="mixhvg-mixture-of-highly-variable-gene-selection-">mixhvg: Mixture of Highly Variable Gene Selection <img src="reference/figures/imgfile.png" align="right" width="160px"><a class="anchor" aria-label="anchor" href="#mixhvg-mixture-of-highly-variable-gene-selection-"></a>
</h1></div>

<div class="section level3">
<h3 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h3>
<p>The package <strong>mixhvg</strong> works for highly variable gene selection, including popular public available methods, and also the mixture of multiple highly variable gene selection methods. The mixture of methods can combine the advantages captured by each single method.</p>
<p>This function <strong>FindVariableFeaturesMix</strong> inherits from <strong>FindVariableFeatures</strong> function of Seurat Package, which can be used the same as FindVariableFeatures. Also, it accepts the dense or sparse matrix input.</p>
<p>Find the tutorial website here <a href="https://mixhvg.github.io" class="external-link uri">https://mixhvg.github.io</a>.</p>
<p>By default, mixhvg accepts a raw count matrix as input and performs the necessary preprocessing internally, including log-normalization and PFlogPF transformation. However, users may also input a Seurat object. If the Seurat object contains only the raw count matrix, mixhvg will handle it the same way as a standalone count matrix. If the Seurat object includes both raw counts and user-processed log-normalized counts, then the log-normalized data in the Seurat object will be used directly for all methods that require it (e.g., lognc methods), while the raw counts will still be used for all other methods (e.g., ct, nc, and PFlogPF methods).</p>
</div>
<div class="section level3">
<h3 id="news">News<a class="anchor" aria-label="anchor" href="#news"></a>
</h3>
<p>Sept. 15th, 2024: The package <code>mixhvg</code> is updated to version 1.0.1, which can support the use in Seurat version 5.</p>
<p>Sept. 1st, 2024: The package <code>mixhvg</code> is updated to version 0.2.1.</p>
<p>Aug. 26th, 2024: Our paper is on biorXiv as a preprint version: <a href="https://doi.org/10.1101/2024.08.25.608519" class="external-link">Zhao, R., Lu, J., Zhou, W., Zhao, N., &amp; Ji, H. (2024). A systematic evaluation of highly variable gene selection methods for single-cell RNA-sequencing. <em>bioRxiv</em>, 2024-08.</a></p>
</div>
<div class="section level3">
<h3 id="download">Download<a class="anchor" aria-label="anchor" href="#download"></a>
</h3>
<p>Please download the <a href="https://CRAN.R-project.org/package=mixhvg" class="external-link"><strong>mixhvg</strong> package from CRAN</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"mixhvg"</span><span class="op">)</span></span></code></pre></div>
<p>Please use GitHub repo to download the most updated package.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"RuzhangZhao/mixhvg"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h3>
<p>There are two inputs can be used in <strong>FindVariableFeaturesMix</strong> function.</p>
<p>The example data comes from <a href="https://www.10xgenomics.com/resources/datasets/pbmc-from-a-healthy-donor-granulocytes-removed-through-cell-sorting-3-k-1-standard-2-0-0" class="external-link">10x Genomics</a>. You may use <a href="https://github.com/RuzhangZhao/pbmc3k/raw/main/pbmc3k_rna.rds" class="external-link">this link</a> to download the processed data. The processed data is named as <code>pbmc3k_rna.rds</code>.</p>
<ul>
<li>Matrix as input: only takes count matrix.</li>
<li>Seurat Object as input:
<ul>
<li>If log normalized result is in object: takes count and normalized matrices from Seurat Object.</li>
<li>If log normalized result is not in object: takes count matrix from Seurat Object.</li>
</ul>
</li>
</ul>
<div class="section level4">
<h4 id="seurat-object-as-input">Seurat Object as Input<a class="anchor" aria-label="anchor" href="#seurat-object-as-input"></a>
</h4>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"pbmc3k_rna.rds"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">mixhvg</span><span class="op">)</span></span>
<span><span class="va">object</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">object</span><span class="op">&lt;-</span><span class="fu"><a href="reference/FindVariableFeaturesMix.html">FindVariableFeaturesMix</a></span><span class="op">(</span><span class="va">object</span>,nfeatures<span class="op">=</span><span class="fl">2000</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># [1] "CD74"  "LYZ"   "IGLC3" "IGKC"  "RPS29" "IGHA1"</span></span></code></pre></div>
<p>One may use <strong>FindVariableFeaturesMix</strong> to replace the <strong>FindVariableFeatures</strong> function in the analysis pipeline of Seurat. For example,</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">object</span><span class="op">&lt;-</span><span class="fu"><a href="reference/FindVariableFeaturesMix.html">FindVariableFeaturesMix</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="va">object</span><span class="op">&lt;-</span><span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">object</span>,verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">object</span><span class="op">&lt;-</span><span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">object</span>,verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">object</span><span class="op">&lt;-</span><span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">object</span>,npcs<span class="op">=</span><span class="fl">30</span>,verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">object</span><span class="op">&lt;-</span><span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">object</span>,dims<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">30</span>,verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="matrix-as-input">Matrix as Input<a class="anchor" aria-label="anchor" href="#matrix-as-input"></a>
</h4>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"pbmc3k_rna.rds"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">mixhvg</span><span class="op">)</span></span>
<span><span class="va">pbmc_hvg</span><span class="op">&lt;-</span><span class="fu"><a href="reference/FindVariableFeaturesMix.html">FindVariableFeaturesMix</a></span><span class="op">(</span><span class="va">pbmc</span>,nfeatures<span class="op">=</span><span class="fl">2000</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pbmc_hvg</span><span class="op">)</span></span>
<span><span class="co"># [1] "CD74"  "LYZ"   "IGLC3" "IGKC"  "RPS29" "IGHA1"</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="different-methods">Different Methods<a class="anchor" aria-label="anchor" href="#different-methods"></a>
</h4>
<p>The <code>method.names</code> can take one method or multiple methods for mixture.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_hvg</span><span class="op">&lt;-</span><span class="fu"><a href="reference/FindVariableFeaturesMix.html">FindVariableFeaturesMix</a></span><span class="op">(</span><span class="va">pbmc</span>,method.names<span class="op">=</span><span class="st">"seuratv3"</span><span class="op">)</span></span>
<span><span class="va">pbmc_hvg</span><span class="op">&lt;-</span><span class="fu"><a href="reference/FindVariableFeaturesMix.html">FindVariableFeaturesMix</a></span><span class="op">(</span><span class="va">pbmc</span>,method.names<span class="op">=</span><span class="st">"scran"</span><span class="op">)</span></span>
<span><span class="va">pbmc_hvg</span><span class="op">&lt;-</span><span class="fu"><a href="reference/FindVariableFeaturesMix.html">FindVariableFeaturesMix</a></span><span class="op">(</span><span class="va">pbmc</span>,</span>
<span>          method.names<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"scran"</span>,<span class="st">"scran_pos"</span>,<span class="st">"seuratv1"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<hr>
<p>In the following example, we use a small cell sorting single-cell RNA-seq dataset with true label, downloaded from a large PBMC dataset. Download the example dataset using <a href="https://github.com/RuzhangZhao/data_pbmc_900" class="external-link">this link</a>.</p>
<p>The data was provided by the work (<a href="https://www.nature.com/articles/s41592-019-0537-1" class="external-link">Wang, J., Agarwal, D., Huang, M., Hu, G., Zhou, Z., Ye, C., &amp; Zhang, N. R. (2019). Data denoising with transfer learning in single-cell transcriptomics. <em>Nature methods</em>, <em>16</em>(9), 875-878.</a>), which was initially downsampled from (<a href="https://www.nature.com/articles/ncomms14049" class="external-link">Zheng, G. X., Terry, J. M., Belgrader, P., Ryvkin, P., Bent, Z. W., Wilson, R., … &amp; Bielas, J. H. (2017). Massively parallel digital transcriptional profiling of single cells. <em>Nature communications</em>, <em>8</em>(1), 14049.</a>).</p>
<p>Load the scRNA-seq data and its corresponding labels.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">expr</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"pbmc_900.rds"</span><span class="op">)</span></span>
<span><span class="va">cell_label</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"cell_labels.txt"</span>,header <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">cell_label</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">cell_label</span><span class="op">$</span><span class="va">cell_type_labels</span>,levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"regulatory_t"</span>,<span class="st">"naive_t"</span>,<span class="st">"cd34"</span>,<span class="st">"memory_t"</span>,<span class="st">"cd56_nk"</span>,<span class="st">"cytotoxic_t"</span>,<span class="st">"cd14_monocytes"</span>,<span class="st">"b_cells"</span>,<span class="st">"naive_cytotoxic"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>There are in total 9 cell type labels, with 100 cells for each cell type.</p>
<p>Here, we check the performance of two different HVG selection methods: Seurat v3 v.s. mixHVG. The evaluation is illustrated by the UMAP plots.</p>
<p>First, we show the standard pipeline for visualizing the dataset with default Seurat pipeline. <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial" class="external-link">Check Seurat tutorial for more details.</a></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">npcs</span><span class="op">=</span><span class="fl">30</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">obj</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span><span class="va">expr</span><span class="op">)</span></span>
<span><span class="va">obj</span><span class="op">$</span><span class="va">cell_label</span><span class="op">&lt;-</span><span class="va">cell_label</span></span>
<span><span class="va">obj</span><span class="op">&lt;-</span><span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">obj</span>,verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">obj</span><span class="op">&lt;-</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">obj</span>,verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">obj</span><span class="op">&lt;-</span><span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">obj</span>,verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="va">obj</span><span class="op">&lt;-</span><span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">obj</span>,npcs<span class="op">=</span><span class="va">npcs</span>,verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">obj</span><span class="op">&lt;-</span><span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">obj</span>,dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="va">npcs</span>,verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/pdf.html" class="external-link">pdf</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"umap_hvg_by_seuratv3.pdf"</span><span class="op">)</span>,height <span class="op">=</span> <span class="fl">5.5</span>,width <span class="op">=</span> <span class="fl">7</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">obj</span>,reduction <span class="op">=</span> <span class="st">"umap"</span>,group.by <span class="op">=</span> <span class="st">"cell_label"</span>,shuffle <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span> </span></code></pre></div>
<p><img src="reference/figures/umap_hvg_by_seuratv3.png" alt="seurat" style="zoom:120%;"></p>
<p>Second, we can simply replace the line <code>FindVariableFeatures</code> by <code>FindVariableFeaturesMix</code> with package mixhvg loaded to apply the mixHVG method.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">mixhvg</span><span class="op">)</span></span>
<span></span>
<span><span class="va">npcs</span><span class="op">=</span><span class="fl">30</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">pbmc</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span><span class="va">expr</span><span class="op">)</span></span>
<span><span class="va">pbmc</span><span class="op">$</span><span class="va">cell_label</span><span class="op">&lt;-</span><span class="va">cell_label</span></span>
<span><span class="va">pbmc</span><span class="op">&lt;-</span><span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">pbmc</span>,verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">pbmc</span><span class="op">&lt;-</span><span class="fu"><a href="reference/FindVariableFeaturesMix.html">FindVariableFeaturesMix</a></span><span class="op">(</span><span class="va">pbmc</span>,verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">pbmc</span><span class="op">&lt;-</span><span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">pbmc</span>,verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">&lt;-</span><span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">pbmc</span>,npcs<span class="op">=</span><span class="va">npcs</span>,verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pbmc</span><span class="op">&lt;-</span><span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">pbmc</span>,dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="va">npcs</span>,verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/pdf.html" class="external-link">pdf</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"umap_hvg_by_mixhvg.pdf"</span><span class="op">)</span>,height <span class="op">=</span> <span class="fl">5.5</span>,width <span class="op">=</span> <span class="fl">7</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">pbmc</span>,reduction <span class="op">=</span> <span class="st">"umap"</span>,group.by <span class="op">=</span> <span class="st">"cell_label"</span>,shuffle <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span> </span></code></pre></div>
<p><img src="reference/figures/umap_hvg_by_mixhvg.png" alt="mixhvg" style="zoom:120%;"></p>
<p>The evaluation is by adjusted rand index (ARI). <a href="https://doi.org/10.1101/2024.08.25.608519" class="external-link">Refer to our paper for more details.</a></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">maxARI</span><span class="op">&lt;-</span><span class="kw">function</span><span class="op">(</span><span class="va">embedding</span>,<span class="va">label</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">label</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">label</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">snn_</span><span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">embedding</span>,</span>
<span>                         nn.method <span class="op">=</span> <span class="st">"rann"</span>,</span>
<span>                         verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span><span class="op">$</span><span class="va">snn</span></span>
<span>    <span class="va">res</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">2</span>,<span class="fl">0.1</span><span class="op">)</span> ,<span class="kw">function</span><span class="op">(</span><span class="va">cur</span><span class="op">)</span><span class="op">{</span></span>
<span>        <span class="va">cluster_label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">snn_</span>,resolution <span class="op">=</span> <span class="va">cur</span>,</span>
<span>                                      verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span>        <span class="fu">mclust</span><span class="fu">::</span><span class="fu"><a href="https://mclust-org.github.io/mclust/reference/adjustedRandIndex.html" class="external-link">adjustedRandIndex</a></span><span class="op">(</span><span class="va">cluster_label</span>,<span class="va">label</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">ari_seuratv3</span><span class="op">&lt;-</span><span class="fu">maxARI</span><span class="op">(</span><span class="va">obj</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">umap</span><span class="op">@</span><span class="va">cell.embeddings</span>,<span class="va">cell_label</span><span class="op">)</span></span>
<span><span class="va">ari_mixhvg</span><span class="op">&lt;-</span><span class="fu">maxARI</span><span class="op">(</span><span class="va">pbmc</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">umap</span><span class="op">@</span><span class="va">cell.embeddings</span>,<span class="va">cell_label</span><span class="op">)</span></span></code></pre></div>
<p>ARI for UMAP with HVG selection by Seurat v3 is <strong>0.6348</strong></p>
<p>ARI for UMAP with HVG selection by mixHVG is <strong>0.6767</strong></p>
<hr>
</div>
</div>
<div class="section level3">
<h3 id="method-choices">Method Choices<a class="anchor" aria-label="anchor" href="#method-choices"></a>
</h3>
<p>The following methods can be chosen. And also, any mixture of the following methods is acceptable. For example, the default is c(“scran”,“seuratv1”,“mv_PFlogPF”,“scran_pos”)</p>
<ul>
<li>
<strong>scran</strong>: Use mean-variance curve adjustment on lognormalized count matrix, which is scran ModelGeneVar.</li>
<li>
<strong>mv_ct</strong>: Use mean-variance curve adjustment on count matrix, inherited from scran ModelGeneVar.</li>
<li>
<strong>mv_nc</strong>: Use mean-variance curve adjustment on normalized count matrix, inherited from scran ModelGeneVar.</li>
<li>
<strong>mv_lognc</strong>: The same as scran.</li>
<li>
<strong>mv_PFlogPF</strong>: Use mean-variance curve adjustment on PFlog1pPF matrix, inherited from scran ModelGeneVar.</li>
<li>
<strong>scran_pos</strong>: Use scran poisson version, modelGeneVarByPoisson.</li>
<li>
<strong>seuratv3</strong>: Use logmean-logvariance curve adjustment on count matrix, which is vst, Seurat FindVariableFeatures Function(<a href="https://satijalab.org/seurat/reference/findvariablefeatures" class="external-link uri">https://satijalab.org/seurat/reference/findvariablefeatures</a>).</li>
<li>
<strong>logmv_ct</strong>: The same as seuratv3.</li>
<li>
<strong>logmv_nc</strong>: Use logmean-logvariance curve adjustment on normalized count matrix, inherited from seuratv3(vst).</li>
<li>
<strong>logmv_lognc</strong>: Use logmean-logvariance curve adjustment on lognormalized count matrix, inherited from seuratv3(vst).</li>
<li>
<strong>logmv_PFlogPF</strong>: Use logmean-logvariance curve adjustment on PFlog1pPF matrix, inherited from seuratv3(vst).</li>
<li>
<strong>seuratv1</strong>: Use dispersion on lognormalized count matrix, which is dispersion (disp), Seurat FindVariableFeatures Function(<a href="https://satijalab.org/seurat/reference/findvariablefeatures" class="external-link uri">https://satijalab.org/seurat/reference/findvariablefeatures</a>).</li>
<li>
<strong>disp_lognc</strong>: The same as seuratv1.</li>
<li>
<strong>disp_PFlogPF</strong>: Use dispersion on PFlog1pPF matrix, inherited from seuratv1(disp).</li>
<li>
<strong>mean_max_ct</strong>: Highly Expressed Features with respect to count matrix.</li>
<li>
<strong>mean_max_nc</strong>: Highly Expressed Features with respect to normalized count matrix.</li>
<li>
<strong>mean_max_lognc</strong>: Highly Expressed Features with respect to lognormalized count matrix</li>
</ul>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=mixhvg" class="external-link">View on CRAN</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing mixhvg</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Ruzhang Zhao <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-3420-5771" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Hongkai Ji <br><small class="roles"> Author </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://cran.r-project.org/package=mixhvg" class="external-link"><img src="http://www.repostatus.org/badges/latest/active.svg" alt="Project Status"></a></li>
<li><a href="https://cran.r-project.org/package=mixhvg" class="external-link"><img src="https://ci.appveyor.com/api/projects/status/a28cs08ug9qng8hn?svg=true" alt="Build app"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Ruzhang Zhao, Hongkai Ji.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
